staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -534,19 +534,16 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 	size := w.endIndex - w.startIndex
 	var oldest uint64
 	switch {
-	case size >= w.capacity:
-		// Once the watch event buffer is full, the oldest watch event we can deliver
-		// is the first one in the buffer.
-		oldest = w.cache[w.startIndex%w.capacity].ResourceVersion
-	case w.listResourceVersion > 0:
-		// If the watch event buffer isn't full, the oldest watch event we can deliver
-		// is one greater than the resource version of the last full list.
+	case w.listResourceVersion > 0 && w.startIndex == 0:
```

Comment:
I don't think `startIndex == 0` means no event was removed from the buffer since relist... startIndex can end up being 0 due to a resize that shrinks the watch cache, right?

```
		// adjust startIndex if cache capacity shrink.
		w.startIndex = w.endIndex - capacity
```

Reply:
this seems to be mixing semantic meaning with what I would have expected to be internal details of a ring buffer...

Reply:
> I don't think `startIndex == 0` means no event was removed from the buffer since relist... startIndex can end up being 0 due to a resize that shrinks the watch cache, right?
> 
> ```
> 		// adjust startIndex if cache capacity shrink.
> 		w.startIndex = w.endIndex - capacity
> ```

startIndex would never be reduced, this increase startIndex by capacity/2 when capacity reduced by half

Reply:
ah, true...